core/src/test/scala/unit/kafka/security/auth/AclTest.scala

```diff
@@ -22,25 +22,29 @@ import kafka.utils.Json
 import org.apache.kafka.common.security.auth.KafkaPrincipal
 import org.junit.{Assert, Test}
 import org.scalatest.junit.JUnitSuite
+
 import scala.collection.JavaConverters._
 
 class AclTest extends JUnitSuite {
 
   val AclJson = "{\"version\": 1, \"acls\": [{\"host\": \"host1\",\"permissionType\": \"Deny\",\"operation\": \"READ\", \"principal\": \"User:alice\"  },  " +
     "{  \"host\":  \"*\" ,  \"permissionType\": \"Allow\",  \"operation\":  \"Read\", \"principal\": \"User:bob\"  },  " +
-    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob\"}  ]}"
+    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob\"},  " +
+    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:\\\\bob\"}, " +
+    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob1\\\\bob2\"}]}"
```

Comment:
These two are still properly escaped strings. We should be testing `User:\\bob` and `"User:bob1\\bob2`? I think this Json would work even without the fallback code.

Reply:
Sure, I tried with the below json: 

` "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:\\bob\"}, " + `
    
`"{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob1\\bob2\"}]}" `

and following set : 

` val acl4 = new Acl(new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "\\bob"), Deny, "host1", Read)`
 
`val acl5 = new Acl(new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "bob1\\bob2"), Deny, "host1", Read) `

Still I don't see fallback code gets hit the execution and throws assertion error for below assert:

 `Assert.assertEquals(acls, Acl.fromBytes(AclJson.getBytes(UTF_8)))  `

`java.lang.AssertionError: 
Expected :Set(User:bob has Deny permission for operations: Read from hosts: host1, User:\bob has Deny permission for operations: Read from hosts: host1, User:bob1\bob2 has Deny permission for operations: Read from hosts: host1, User:bob has Allow permission for operations: Read from hosts: *, User:alice has Deny permission for operations: Read from hosts: host1)`

`Actual   :Set(User:bob1|bob2 has Deny permission for operations: Read from hosts: host1, User:bob has Deny permission for operations: Read from hosts: host1, User:|bob has Deny permission for operations: Read from hosts: host1, User:bob has Allow permission for operations: Read from hosts: *, User:alice has Deny permission for operations: Read from hosts: host1)`

I noticed that `Acl.fromBytes(jsonAcls)` has different values than `Acl.fromBytes(AclJson.getBytes(UTF_8))`. However completely not sure the possible reasons behind this.

Reply:
@rajinisivaram, a gentle reminder on above message.

Reply:
I recommend using `"""` to make it easier to understand what is being escaped.

Reply:
@ijuma I tried using `"""`. Here is my test string: 
```
val AclJson = """{"version": 1, "acls": [{"host": "host1","permissionType": "Deny","operation": "READ", "principal": "User:alice"  },
    {  "host":  "*" ,  "permissionType": "Allow",  "operation":  "Read", "principal": "User:bob"  },
    {  "host": "host1",  "permissionType": "Deny",  "operation":   "Read" ,  "principal": "User:bob"},
    {  "host": "host1",  "permissionType": "Deny",  "operation":   "Read" ,  "principal": "User:\\bob"},
    {  "host": "host1",  "permissionType": "Deny",  "operation":   "Read" ,  "principal": "User:bob1\\bob2"}]}"""
```

And this string also doesn't hit the fall back code. It simply gets parsed without error and the escaped principals return back as below: 

```
User:\bob has Deny permission for operations: Read from hosts: host1
User:bob1\bob2 has Deny permission for operations: Read from hosts: host1 
```
Not completely sure about this behavior though. Any suggestions would be helpful.